### DISCUSSION TITLE
Robot penetrate the terrain or get stuck in gaps

### DISCUSSION THREAD

**Discussion by zitongbai** [google-deepmind/mujoco#2307] [Asking for Help] [answered]
### Intro

Hi!

I am a graduate student, I use MuJoCo for my research on legged robot.

### My setup

MuJoCo version 3.2.6

### My question

I simulated robot walking on the terrain created by geoms (like `<geom pos="8.2 0.0 0.15" type="box" size="0.1 0.75 0.15"
      quat="1.0 0.0 0.0 0.0" />`). But I found that robot's leg may penetrate through the geom and it would be stuck in the gap between geoms and ground or between two geoms. Is there any method to resolve this phenomenon? 

![Screenshot from 2024-12-23 00-50-31](https://github.com/user-attachments/assets/4e011a4a-8bbd-488e-85d8-8597036b6546)

![Screenshot from 2024-12-23 00-51-13](https://github.com/user-attachments/assets/022a92dd-59f5-4f4a-be9e-10cfd02729f8)

![Screenshot from 2024-12-23 00-57-39](https://github.com/user-attachments/assets/e652ccf2-ea3a-4cfb-ba8b-635ef48d1ac1)

![Screenshot from 2024-12-23 00-58-11](https://github.com/user-attachments/assets/19bfc216-17c5-4611-a1c8-49d40ec12595)

Thanks for your attention and help in advance. 

### Minimal model and/or code that explain my question

Model:

```XML

<mujoco model="go2 scene">
  <include file="go2.xml" />

  <statistic center="0 0 0.1" extent="0.8" />

  <visual>
    <headlight diffuse="0.6 0.6 0.6" ambient="0.3 0.3 0.3" specular="0 0 0" />
    <rgba haze="0.15 0.25 0.35 1" />
    <global azimuth="-130" elevation="-20" />
  </visual>

  <asset>
    <texture type="skybox" builtin="gradient" rgb1="0.3 0.5 0.7" rgb2="0 0 0" width="512"
      height="3072" />
    <texture type="2d" name="groundplane" builtin="checker" mark="edge" rgb1="0.2 0.3 0.4"
      rgb2="0.1 0.2 0.3" markrgb="0.8 0.8 0.8" width="300" height="300" />
    <material name="groundplane" texture="groundplane" texuniform="true" texrepeat="5 5"
      reflectance="0.2" />
    <hfield name="perlin_hfield" size="1.0 0.75 0.2 0.2" file="../height_field.png" />
    <hfield name="image_hfield" size="1.0 1.0 0.02 0.1" file="../unitree_hfield.png" />
  </asset>

  <worldbody>
    <light pos="0 0 1.5" dir="0 0 -1" directional="true" />
    <geom name="floor" size="0 0 0.05" type="plane" material="groundplane" />
    <geom pos="2.0 0.0 0.4" type="box" size="1.5 0.75 0.05"
      quat="0.984807753012208 0.0 -0.17364817766693033 0.0" />
    <geom pos="4.78 0.0 0.4" type="box" size="1.5 0.75 0.05"
      quat="0.984807753012208 0.0 0.17364817766693033 0.0" />
    <geom pos="7.2 0.0 0.025" type="box" size="0.10500000000000001 0.75 0.025"
      quat="1.0 0.0 0.0 0.0" />
    <geom pos="7.4 0.0 0.05" type="box" size="0.10500000000000001 0.75 0.05" quat="1.0 0.0 0.0 0.0" />
    <geom pos="7.6 0.0 0.07500000000000001" type="box"
      size="0.10500000000000001 0.75 0.07500000000000001" quat="1.0 0.0 0.0 0.0" />

<!-- something else -->

  </worldbody>
</mujoco>

```

Detailed model file can be found [here](https://github.com/zitongbai/legged_rl/blob/main/legged_robot_example/unitree_go2_description/mjcf/scene_terrain_1.xml) 

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by yuvaltassa** [comment, marked as answer]
The easiest solution is to make the static geoms overlap.

**Reply by zitongbai** [reply]
Thanks for your suggestion, I will take a try.

**Reply by zitongbai** [reply]
It do work, but I'm curious if there are more sophisticated approaches available. Could you recommend some documentation for further reference?

**Reply by yuvaltassa** [reply]
You could add collision margins. Search for "margin" in the docs. But margins are invisible so are usually not what you want.
